File: audiolib/src/main/cpp/native/AudioNative.cpp
#include "AudioNative.h"
#include "audio/SilkCoder.h"
#include "audio/LameCoder.h"

void silkToPcmConvert(
        JNIEnv *env, jobject clazz,
        jstring silkPath, jstring pcmPath, jboolean debug,
        jint fs_Hz, jint loss
) {
    const char *nativeSilkPath = env->GetStringUTFChars(silkPath, nullptr);
    const char *nativePcmPath = env->GetStringUTFChars(pcmPath, nullptr);
    SilkCoder::SilkToPcmConvert(
            nativeSilkPath, nativePcmPath, debug,
            fs_Hz, loss
    );
    env->ReleaseStringUTFChars(silkPath, nativeSilkPath);
    env->ReleaseStringUTFChars(pcmPath, nativePcmPath);
}

void pcmToSilkConvert(
        JNIEnv *env, jobject clazz,
        jstring pcmPath, jstring silkPath, jboolean tencent, jboolean debug,
        jint fs_Hz, jint maxInternalSampleRate, jint packetSizeMs, jint packetLossPercentage,
        jint useInBandFEC, jint useDTX, jint complexity, jint bitRate
) {
    const char *nativePcmPath = env->GetStringUTFChars(pcmPath, nullptr);
    const char *nativeSilkPath = env->GetStringUTFChars(silkPath, nullptr);
    SilkCoder::PcmToSilkConvert(
            nativePcmPath, nativeSilkPath, tencent, debug,
            fs_Hz, maxInternalSampleRate, packetSizeMs, packetLossPercentage,
            useInBandFEC, useDTX, complexity, bitRate
    );
    env->ReleaseStringUTFChars(pcmPath, nativePcmPath);
    env->ReleaseStringUTFChars(silkPath, nativeSilkPath);
}

void pcmToMp3Init(
        JNIEnv *env, jobject clazz,
        jint inputSampleRate, jfloat inputScale, jint inputChannelCount,
        jint outputSampleRate, jint outputBitrate,
        jint encodingQuality, jint vbrQuality, jint abrMeanBitrate,
        jint vbrMode, jint mode,
        jint lowPassFrequency, jint highPassFrequency
) {
    LameCoder::PcmToMp3Init(
            inputSampleRate, inputScale, inputChannelCount,
            outputSampleRate, outputBitrate,
            encodingQuality, vbrQuality, abrMeanBitrate,
            vbrMode, mode,
            lowPassFrequency, highPassFrequency
    );
}

void pcmToMp3Convert(
        JNIEnv *env, jobject clazz,
        jstring pcmPath, jstring mp3Path
) {
    const char *nativePcmPath = env->GetStringUTFChars(pcmPath, nullptr);
    const char *nativeMp3Path = env->GetStringUTFChars(mp3Path, nullptr);
    LameCoder::PcmToMp3Convert(
            nativePcmPath, nativeMp3Path
    );
    env->ReleaseStringUTFChars(pcmPath, nativePcmPath);
    env->ReleaseStringUTFChars(mp3Path, nativeMp3Path);
}

void pcmToMp3Close(
        JNIEnv *env, jobject clazz
) {
    LameCoder::PcmToMp3Close();
}

void mp3ToPcmInit(
        JNIEnv *env, jobject clazz
) {
    LameCoder::Mp3ToPcmInit();
}

int mp3ToPcmConvert(
        JNIEnv *env, jobject clazz,
        jstring mp3Path, jstring pcmPath
) {
    const char *nativeMp3Path = env->GetStringUTFChars(mp3Path, nullptr);
    const char *nativePcmPath = env->GetStringUTFChars(pcmPath, nullptr);
    int sampleRate = LameCoder::Mp3ToPcmConvert(
            nativeMp3Path, nativePcmPath
    );
    env->ReleaseStringUTFChars(mp3Path, nativeMp3Path);
    env->ReleaseStringUTFChars(pcmPath, nativePcmPath);
    return sampleRate;
}

void mp3ToPcmClose(
        JNIEnv *env, jobject clazz
) {
    LameCoder::Mp3ToPcmClose();
}

void AudioNative::init(JNIEnv *env) {
    jclass clazz = env->FindClass("me/hd/audiolib/native/AudioNative");
    JNINativeMethod methods[] = {
            {"silkToPcmConvert", "(Ljava/lang/String;Ljava/lang/String;ZII)V",        (void *) silkToPcmConvert},
            {"pcmToSilkConvert", "(Ljava/lang/String;Ljava/lang/String;ZZIIIIIIII)V", (void *) pcmToSilkConvert},
            {"pcmToMp3Init",     "(IFIIIIIIIIII)V",                                   (void *) pcmToMp3Init},
            {"pcmToMp3Convert",  "(Ljava/lang/String;Ljava/lang/String;)V",           (void *) pcmToMp3Convert},
            {"pcmToMp3Close",    "()V",                                               (void *) pcmToMp3Close},
            {"mp3ToPcmInit",     "()V",                                               (void *) mp3ToPcmInit},
            {"mp3ToPcmConvert",  "(Ljava/lang/String;Ljava/lang/String;)I",           (void *) mp3ToPcmConvert},
            {"mp3ToPcmClose",    "()V",                                               (void *) mp3ToPcmClose},
    };
    const int methodCount = sizeof(methods) / sizeof(methods[0]);
    if (env->RegisterNatives(clazz, methods, methodCount) < 0) {
        LOGE("AudioNative registerNatives failed");
    }
    env->DeleteLocalRef(clazz);
}

File: audiolib/src/main/java/me/hd/audioconverter/native/AudioNative.kt
package me.hd.audioconverter.native

object AudioNative {
    external fun silkToPcmConvert(
        silkPath: String,
        pcmPath: String,
        debug: Boolean = false,
        fs_Hz: Int = 24000,
        loss: Int = 0
    )

    external fun pcmToSilkConvert(
        pcmPath: String,
        silkPath: String,
        tencent: Boolean = true,
        debug: Boolean = false,
        fs_Hz: Int = 24000,
        maxInternalSampleRate: Int = 24000,
        packetSizeMs: Int = 20,
        packetLossPercentage: Int = 0,
        useInBandFEC: Int = 0,
        useDTX: Int = 0,
        complexity: Int = 2,
        bitRate: Int = 24000
    )

    external fun pcmToMp3Init(
        inputSampleRate: Int = 24000,
        inputScale: Float = 1f,
        outputChannelCount: Int = 1,
        outputSampleRate: Int = 24000,
        outputBitrate: Int = 48,
        encodingQuality: Int = 8,
        vbrQuality: Int = 5,
        abrMeanBitrate: Int = 128,
        vbrMode: Int = 0,
        mode: Int = 3,
        lowPassFrequency: Int = 0,
        highPassFrequency: Int = 0
    )

    external fun pcmToMp3Convert(
        pcmPath: String,
        mp3Path: String
    )

    external fun pcmToMp3Close()

    external fun mp3ToPcmInit()

    external fun mp3ToPcmConvert(
        mp3Path: String,
        pcmPath: String
    ): Int

    external fun mp3ToPcmClose()

    fun silkToPcm(silkPath: String, pcmPath: String) {
        silkToPcmConvert(silkPath, pcmPath)
    }

    fun pcmToSilk(pcmPath: String, silkPath: String, sampleRate: Int = 24000) {
        pcmToSilkConvert(pcmPath, silkPath, fs_Hz = sampleRate)
    }

    fun pcmToMp3(pcmPath: String, mp3Path: String) {
        pcmToMp3Init()
        pcmToMp3Convert(pcmPath, mp3Path)
        pcmToMp3Close()
    }

    fun mp3ToPcm(mp3Path: String, pcmPath: String): Int {
        mp3ToPcmInit()
        val sampleRate = mp3ToPcmConvert(mp3Path, pcmPath)
        mp3ToPcmClose()
        return sampleRate
    }
}